### DISCUSSION TITLE
Attempting to use renovate to do a phased rollout of new dependencies

### DISCUSSION THREAD

**Discussion by youngm** [renovatebot/renovate#21462] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I'd like to use renovate to do a phased rollout of a new version of a dependency. I have hundreds of repositories that depend upon this dependency and would like to making so that they upgrade in phases depending upon the SLA of the project.

### What you tried so far.

My thought is to use `stabilityDays` to set the phase the repository will pick up the dependency and combine it with `automerge=true` so that after the stability days have passed the dependency upgrade will be merged and released.

I have hundreds of repositories that need to be updated and I don't want to overload my CI/CD system when the new dependency is released. To help with this I have set a PR to not be created until it is "not-pending". So, I have set `prCreation=not-pending` and my CI/CD is set to only build on changes/creation of PRs not branches.

I want to be sure that the PR CI passes before the PR is automerged so I have set `internalChecksAsSuccess=false`.

Looking at the logs, it appears that configuration does not work. A PR isn't getting created because "not-pending" won't allow the PR to be created since stability days is the only passing check.

I imagine I could fix this by setting prCreation=immediate or enable builds on branches. But, I'm afraid enabling both of those actions would DDOS my CI/CD system with PR check builds.

I could also set `internalChecksAsSuccess=true`. But, in the past I've had PRs automerged before the CI/CD has completed when that setting is true.

Anyone have any ideas or thoughts?

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: syncBranchState()(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: branch.isUpToDate(): using cached result "true"(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: getBranchPr(renovate/com.azure-azure-messaging-eventgrid-4.x)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: findPr(renovate/com.azure-azure-messaging-eventgrid-4.x, undefined, open)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: findPr(renovate/com.azure-azure-messaging-eventgrid-4.x, undefined, closed)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: branchExists=true(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: dependencyDashboardCheck=undefined(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: recreateClosed is false(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: findPr(renovate/com.azure-azure-messaging-eventgrid-4.x, Update dependency com.azure:azure-messaging-eventgrid to v4.15.0, !open)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: prAlreadyExisted=false(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Checking if PR has been edited(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: branch.isModified(): using cached result "false"(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Checking schedule(after 10pm every weekday,before 5am every weekday,every weekend, America/Denver)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Found timezone: America/Denver(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Adjusting now for timezone(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Checking 3 schedule(s)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Checking schedule "after 10pm every weekday"(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
{
  "parsedSchedule": {
    "schedules": [
      {
        "t_a": [
          79200
        ],
        "d": [
          2,
          3,
          4,
          5,
          6
        ]
      }
    ],
    "exceptions": [],
    "error": -1
  }
}
DEBUG: Matches schedule after 10pm every weekday(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Branch already exists(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: getBranchPr(renovate/com.azure-azure-messaging-eventgrid-4.x)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: findPr(renovate/com.azure-azure-messaging-eventgrid-4.x, undefined, open)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: findPr(renovate/com.azure-azure-messaging-eventgrid-4.x, undefined, closed)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: branch.isBehindBase(): using cached result "false"(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Branch is up-to-date(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: isBranchConflicted(main, renovate/com.azure-azure-messaging-eventgrid-4.x)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: branch.isConflicted(): using cached result "false"(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Branch does not need rebasing(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Using reuseExistingBranch: true(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Status check renovate/stability-days is already up-to-date(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Checking if we can automerge branch(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: mergeStatus=no automerge(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Ensuring PR(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: There are 0 errors and 0 warnings(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: getBranchPr(renovate/com.azure-azure-messaging-eventgrid-4.x)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: findPr(renovate/com.azure-azure-messaging-eventgrid-4.x, undefined, open)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: findPr(renovate/com.azure-azure-messaging-eventgrid-4.x, undefined, closed)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: getPrCache()(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Checking branch combined status(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: resolveBranchStatus(branchName=renovate/com.azure-azure-messaging-eventgrid-4.x, ignoreTests=false)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: getBranchStatus(renovate/com.azure-azure-messaging-eventgrid-4.x)(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: branch status check result(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
{
  "state": "success",
  "statuses": [
    {
      "url": "https://api.github.com/repos/ICSEng/emx-core/statuses/91e931d5aaefc488c6e7d38f7640e605c0a1af26",
      "avatar_url": "https://avatars.githubusercontent.com/in/2740?v=4",
      "id": 22513701998,
      "node_id": "SC_kwDOFggxy88AAAAFPevUbg",
      "state": "success",
      "description": "Updates have met stability days requirement",
      "target_url": "https://docs.renovatebot.com/",
      "context": "renovate/stability-days",
      "created_at": "2023-04-11T05:23:02Z",
      "updated_at": "2023-04-11T05:23:02Z"
    }
  ]
}
DEBUG: Successful checks are all internal renovate/ checks, so returning "pending" branch status(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: No check runs found(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
{
  "result": {
    "total_count": 0,
    "check_runs": []
  }
}
DEBUG: Branch status yellow(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Branch status is yellow - checking timeout(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
DEBUG: Branch is 23 hours old - skipping PR creation(branch="renovate/com.azure-azure-messaging-eventgrid-4.x")
```

</details>

**Comment by youngm** [comment]
I really had high hopes for `internalChecksAsSuccess`. If that flag only applied to automerge decisions and not PR creation then I think I'd be golden.

**Reply by rarkins** [reply]
I also thought that flag would apply to automerge too. Can you find the log messages for an applicable branch at the time when it "should" automerge? e.g. the logs might say skipping due to no tests

**Reply by youngm** [reply]
Hello rarkins,

It does apply to automerge. The problem for me is that it is also applied to PR creation checks. Because my CI doesn't create a build until there is a PR, using `internalChecksAsSuccess=false` and `prCreation=not-pending` together result PRs are never created.

**Reply by viceice** [reply]
don't use not-pending when running CI checks only on PR

**Reply by youngm** [reply, marked as answer]
Thanks viceice. Yeah, that's what I'm doing now in a limited manor. However, like I mentioned in my original post, when there is a new release I'd like to avoid DDOSing my CI/CD agents with hundreds of new PR builds. We don't have an unlimited scale CI/CD system. I'd prefer for them to phase in. I was hoping stabilityDays+prCreation=not-pending would allows PRs to trickle in rather than flood in. Any thoughts on how I could perhaps roll in the PR creation for my hundreds of projects with prCreation=immediate?

**Reply by youngm** [reply]
Hmm....I wonder if I could use `prNotPendingHours` instead of stabilityDays to phase PR creation?

**Reply by viceice** [reply]
use internal filter strict, then Renovate will suppress branch (and PR) untill stability is passing

**Reply by youngm** [reply]
@viceice can you help me understand what other config you might set for this idea of setting `internalChecksFilter=strict` and stabilityDays?

For the record I did have a solution working previously with `internalChecksAsSuccess=true`. But, I ran into trouble with the PR being auto merged before the PR build was run. That is why this go around I'm trying to base the config on `internalChecksAsSuccess=false`.

**Reply by viceice** [reply]
leave `internalChecksFilter` (strict), `internalChecksAsSuccess` (false) and `prCreation` (immediate) as default. set `minimumReleaseAge` (previously stabillityDays)  to an apropriate value. Let renovate do (should create branch an pr after `minimumReleaseAge` is passed). Delete existing branches before, otherwise renovate will immediately create the PR.

**Remember** Not all datasources and registries return timestamps or they return wrong timestamps. So `minimumReleaseAge` does nothing. 

Use required status check on branch protections to not merge before checks are passing.

**Reply by youngm** [reply]
Thank you @viceice. I hadn't considered using "Require status checks" because it could be difficult to maintain across my 300 some odd repositories. But, that could work. I'll investigate that option further.

Can you think of any other possible configuration combinations for me?

**Reply by viceice** [reply]
you can probably use the new organization rules

**Reply by youngm** [reply]
Wait, as I look closer at your suggestion I'm confused. Did you mean internalChecksAsSuccess=true and prCreation=non-pending? prCreation=immediate it seems could DDOS my CI/CD which I'm trying to avoid. And internalChecksAsSuccess=false causes prCreation=non-pending to not work. Can you help me understand how the combination you suggested would work in theory?

I can see how internalChecksAsSuccess=true and prCreation=non-pending with status check on branch could work.

**Reply by viceice** [reply]
no, do as i told before. leave everything as default and add minimum release age.

then renovate will suppress the PR untill age is reached. you need s different age for each repo.

you can now add branch protections to GitHub organization.

**Reply by youngm** [reply]
@viceice Thanks for your help so far. Unfortunately I am not currently able to enable "Require status check". I have some pipelines that commit changes to main and I'm unable to allow my pipeline to commit to main when "Require status check" is set. :(

I'm starting to see how my current pipeline situation is really getting in the way of this problem and not Renovate. Do you have any other ideas for me? If not I'll go ahead and mark this as solved and I'll see if I can break loose one of these CI constraints I'm working under.

**Comment by youngm** [comment]
So, I've left the settings recommended by @viceice for a few weeks now and it appears all of the upgrade PRs successfully build before they are auto merged. Is that just a matter of luck? Or is this functionality I can rely upon?

**Reply by viceice** [reply]
that should be the way renovate works
